File: src/main/scala/com/data/etl/WordCount.scala
package com.data.etl

import java.io.{StringReader, StringWriter}

import au.com.bytecode.opencsv.{CSVReader, CSVWriter}
import org.apache.hadoop.io.{IntWritable, Text}
import org.apache.hadoop.mapreduce.Job
import org.apache.hadoop.mapreduce.lib.input.KeyValueTextInputFormat
import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat
import org.apache.spark.sql.SparkSession
import org.apache.spark.{SparkConf, SparkContext}
import org.json4s.jackson.JsonMethods._
import org.json4s.{DefaultFormats, _}

import scala.collection.JavaConversions._


case class Person(firstName:String, lastName:String, address:List[Address]){
  override def toString = s"Person(firstName=$firstName, lastName=$lastName, address=$address"
}
case class Address(line1:String, city:String, state:String, zip:String) {
  override def toString = s"Address(line1=$line1, city=$city, state=$state, zip=$zip)"
}

// 读取JSON
object WordCount {

  def main(args: Array[String]): Unit = {
    val inputJsonFile = "file:///home/wordcount.json"
    val conf = new SparkConf().setAppName("wordCount").setMaster("local")
    val sc = new SparkContext(conf)
    val input5 = sc.textFile(inputJsonFile)
    val dataObjsRDD = input5.map{ myrecord =>
      implicit val formats = DefaultFormats

      val jsonObj = parse(myrecord)
      jsonObj.extract[Person]
    }
    dataObjsRDD.saveAsTextFile("file:///home/word.json")

  }
}
//读取CSV文件
object WordCount2 {
  def main(args: Array[String]) {

    val conf = new SparkConf().setAppName("wordCount").setMaster("local")
    val sc = new SparkContext(conf)
    val input = sc.textFile("/sample_map.csv")
    val result6 = input.map{ line =>
      val reader = new CSVReader(new StringReader(line));
      reader.readNext();
    }
    for(result <- result6){
      for(re <- result){
        println(re)
      }
    }
  }
}

case class Data(index: String, title:String, content:String)

object WordCount3 {
  def main(args: Array[String]): Unit = {

    val conf = new SparkConf().setAppName("wordCount").setMaster("local")
    val sc = new SparkContext(conf)
    val input = sc.wholeTextFiles("d:/sample_map.csv")

    val result = input.flatMap{ case (_, txt) =>
    val reader = new CSVReader(new StringReader(txt));
//      reader.readAll().map( x=> Data(x(0), x(1), x(2)))
      reader.readAll()
    }

    result.collect().foreach( x=> {
      x.foreach(println); println("=======")
    })
  }
}

// 保存CSV
object WordCount4 {
  def main(args: Array[String]): Unit = {

    val conf = new SparkConf().setAppName("WordCount").setMaster("local")
    val sc = new SparkContext(conf)
    val inputRDD = sc.parallelize(List(Data("index", "title", "content")))
    inputRDD.map(data => List(data.index, data.title, data.content).toArray)
      .mapPartitions { data =>
        val stringWriter = new StringWriter();
        val csvWriter = new CSVWriter(stringWriter);
        csvWriter.writeAll(data.toList)
        Iterator(stringWriter.toString)
      }.saveAsTextFile("/home/common/coding/coding/Scala/word-count/sample_map_out")
  }
}

object SparkRDD {

  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName("Wordc").setMaster("local")
    val sc = new SparkContext(conf)

    //写sequenceFiles
    val rdd = sc.parallelize(List(("pandas", 3), ("kay", 6),("Snail", 2)))
    rdd.saveAsSequenceFile("output")

    //读取sequenceFile
    val output = sc.sequenceFile("output", classOf[Text], classOf[IntWritable]).
      map{ case (x, y) => (x.toString, y.get())}
    output.foreach(println)
  }
}

object SparkRDD2 {

  def main(args: Array[String]) {
    val conf = new SparkConf().setAppName("WordCount").setMaster("local")
    val sc = new SparkContext(conf)

    //使用老式 API 读取 KeyValueTextInputFormat()，以JSON文件为例子
    //注意引用的包是org.apache.hadoop.mapred.KeyValueTextInputFormat
    //    val input = sc.hadoopFile[Text, Text, KeyValueTextInputFormat]("input/test.json").map {
    //      case (x, y) => (x.toString, y.toString)
    //    }
    //    input.foreach(println)

    // 读取文件，使用新的API，注意引用的包是org.apache.hadoop.mapreduce.lib.input.KeyValueTextInputFormat
    val job = new Job()
    val data = sc.newAPIHadoopFile("input/test.json",
      classOf[KeyValueTextInputFormat],
      classOf[Text],
      classOf[Text],
      job.getConfiguration)
    data.foreach(println)

    //保存文件，注意引用的包是org.apache.hadoop.mapreduce.lib.output.TextOutputFormat
    data.saveAsNewAPIHadoopFile(
      "input/test1.json",
      classOf[Text],
      classOf[Text],
      classOf[TextOutputFormat[Text, Text]],
      job.getConfiguration)

  }
}

  object WordCount5 {
    def main(args: Array[String]): Unit = {

      val conf = new SparkConf().setAppName("Wc5").setMaster("local")
      val spark = SparkSession.builder().config(conf).enableHiveSupport().getOrCreate()
      val sc = spark.sparkContext

      sc.setLogLevel("DEBUG")
      val txtfile = sc.textFile("D:\\code\\java\\sparksqltrain\\src\\main\\scala\\com\\data\\etl\\readme.md").flatMap(_.split(" "))
        .map(word => (word, 1)).reduceByKey(_+_, 1)
        .map(pair => (pair._2, pair._1))
        .sortByKey(false)
        .map(pair => (pair._2, pair._1)).cache()
        //.collect()
      txtfile.foreach(x => println(x))

    }
  }

File: src/main/java/com/data/service/WordCount.java
package com.data.service;

import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;

import scala.Tuple2;


public class WordCount {
    public static void main(String[] args) {
        System.out.println(args[0]);
        System.out.println(args[1]);
        System.out.println(args[2]);
        //1.本地模式，创建spark配置及上下文
        //	SparkConf conf = new SparkConf().setAppName("wordCount").setMaster("spark://192.168.174.132:7077");
        SparkConf conf = new SparkConf().setAppName("wordCount").setMaster(args[0]);

        JavaSparkContext sc = new JavaSparkContext(conf);

        //2.读取本地文件,并创建RDD
        //	JavaRDD<String> linesRDD = sc.textFile("e:\\words.txt");
        //	JavaRDD<String> linesRDD = sc.textFile("hdfs://192.168.174.132:9000/wc_input");
        JavaRDD<String> linesRDD = sc.textFile(args[1]);
        //3.每个单词由空格隔开,将每行的linesRDD拆分为每个单词的RDD
        //	JavaRDD<String> wordsRDD = linesRDD.flatMap(s  -> Arrays.asList(s.split("\\s")));
        //相当于 ==>
        JavaRDD<String> wordsRDD = linesRDD.flatMap(new FlatMapFunction<String, String>(){
            private static final long serialVersionUID = 1L;
            @SuppressWarnings("unchecked")
            @Override
            public Iterator<String> call(String line) throws Exception {
                return Arrays.asList(line.split(" ")).listIterator();
            }
        });
        //4.将每个单词转为key-value的RDD，并给每个单词计数为1
        //JavaPairRDD<String,Integer> wordsPairRDD = wordsRDD.mapToPair(s -> new Tuple2<String,Integer>(s, 1));
        //相当于 ==>
        JavaPairRDD<String,Integer> wordsPairRDD = wordsRDD.mapToPair(new PairFunction<String, String, Integer>() {
            private static final long serialVersionUID = 1L;
            @Override
            public Tuple2<String, Integer> call(String word) throws Exception {
                return new Tuple2<String,Integer>(word,1);
            }
        });

        //5.计算每个单词出现的次数
        //JavaPairRDD<String,Integer> wordsCountRDD = wordsPairRDD.reduceByKey((a,b) -> a+b);
        //相当于 ==>
        JavaPairRDD<String,Integer> wordsCountRDD = wordsPairRDD.reduceByKey(new Function2<Integer, Integer, Integer>() {
            private static final long serialVersionUID = 1L;

            @Override
            public Integer call(Integer v1, Integer v2) throws Exception {
                return v1 + v2;
            }
        });

        //6.因为只能对key进行排序，所以需要将wordsCountRDD进行key-value倒置，返回新的RDD
        // JavaPairRDD<Integer,String> wordsCountRDD2 = wordsCountRDD.mapToPair(s -> new Tuple2<Integer,String>(s._2, s._1));
        //相当于 ==>
        JavaPairRDD<Integer,String> wordsCountRDD2 = wordsCountRDD.mapToPair(new PairFunction<Tuple2<String,Integer>, Integer, String>() {
            private static final long serialVersionUID = 1L;
            @Override
            public Tuple2<Integer, String> call(Tuple2<String, Integer> t) throws Exception {
                return new Tuple2<Integer,String>(t._2,t._1);
            }
        });

        //7.对wordsCountRDD2进行排序,降序desc
        JavaPairRDD<Integer,String> wordsCountRDD3 = wordsCountRDD2.sortByKey(false);

        //8.只取前10个
        List<Tuple2<Integer, String>>  result = wordsCountRDD3.take(10);

        //9.打印
        //	 result.forEach(t -> System.out.println(t._2 + "   " + t._1));
             for(Tuple2<Integer, String> t : result){
            System.out.println(t._2 + "   " + t._1);
        }

        //10.将结果保存到文件中
             wordsCountRDD3.saveAsTextFile(args[2]);
             sc.close();
        }
}
